### PR TITLE
Faster TIS analysis

### PR DESCRIPTION
**Pull request by dwhswenson**
Just a few lines of code to provide a significant speed-up for TIS analysis.

Caching previous results of max_lambda, so we don't check it again if we have the same trajectory as before. For the toy model, this nearly doubles the speed (with 20k trajectories).

Before:

image

After:

image

Same results; we just get them faster.

### CODE REVIEW

**Comment by dwhswenson** [comment]
Passes tests; ready for review and merge.

I'll be doing other stuff in other PRs to make this even faster (and when #662 is done, that will make it possible to do other tricks to speed it up.)

The speed-up from this PR will be better with more ensembles or with lower acceptance ratios.

Interestingly, profiling the rate matrix calculation now says that, for 20k steps in the toy MSTIS example, the main cost is in identifying which steps had minus moves (and therefore need to be included in the flux calculation.) I think I can find a way to fix that. (Currently, it compares the ensembles; ensemble equality is checked based on the string representation of the ensemble; making those strings actually takes 30% of the total analysis time!)

**Comment by jhprinz** [comment]
Looks good, neat trick. There is one line I think can be moved for probably unnoticeable additional improvement :)

File: openpathsampling/high_level/transition.py [hunk 1]
```diff
@@ -257,10 +257,18 @@ def ensemble_statistics(self, ensemble, samples, weights=None, force=False):
         hist_data = {}
         buflen = -1
         sample_buf = []
+        prev_sample = {h: None for h in run_it}
+        prev_result = {h: None for h in run_it}
         for sample in in_ens_samples:
             for hist in run_it:
                 hist_info = self.ensemble_histogram_info[hist]
```
Comment (jhprinz) [inline openpathsampling/high_level/transition.py]: I would move this in the else sub clause. I think it is only called when a new sample is found correct?

Comment (dwhswenson) [inline openpathsampling/high_level/transition.py]: True -- but I'll add that to #666. You might as well just merge that one, which will auto-merge this as well.

**Pull request merged**